### PR TITLE
RGB<->YIQ colorspace conversion

### PR DESCRIPTION
**Pull request by mystic123**
Implementation of functions for colorspace conversion RGB <-> YIQ, according to [https://en.wikipedia.org/wiki/YIQ#Formulas](Wikipedia).
This PR adds CPP functions and python wrappers in tf.image namespace

### CODE REVIEW

**Comment by yifeif** [comment]
Thanks @mystic123! @martinwicke would you be able to review this PR?

File: tensorflow/core/kernels/colorspace_op.h [hunk 1]
```diff
@@ -89,6 +89,45 @@ struct HSVToRGB {
   }
 };
 
+// following algorithm from https://en.wikipedia.org/wiki/YIQ#Formulas
```
Comment (martinwicke) [inline tensorflow/core/kernels/colorspace_op.h]: Nit: Please capitalize "Following"

File: tensorflow/python/ops/image_ops_impl.py [hunk 2]
```diff
@@ -45,6 +45,8 @@
 # TODO(b/31222613): This op may be differentiable, and there may be
 # latent bugs here.
 ops.NotDifferentiable('HSVToRGB')
+ops.NotDifferentiable('YIQToRGB')
+ops.NotDifferentiable('RGBToYIQ')
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_impl.py]: These look pretty differentiable to me... it's just a linear transformation, no?

File: tensorflow/python/ops/image_ops_test.py [hunk 3]
```diff
@@ -84,6 +84,43 @@ def testRGBToHSVRoundTrip(self):
...
+      inp = np.random.rand(*shape).astype(nptype)
+
+      # Convert to YIQ and back, as a batch and individually
+      with self.test_session(use_gpu=True) as sess:
+        batch0 = constant_op.constant(inp)
+        batch1 = image_ops.rgb_to_yiq(batch0)
+        batch2 = image_ops.yiq_to_rgb(batch1)
+        split0 = array_ops.unstack(batch0)
+        split1 = list(map(image_ops.rgb_to_hsv, split0))
+        split2 = list(map(image_ops.hsv_to_rgb, split1))
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_test.py]: Why hsv here?

File: tensorflow/python/ops/image_ops_test.py [hunk 4]
```diff
@@ -84,6 +84,43 @@ def testRGBToHSVRoundTrip(self):
...
+      self.assertAllClose(batch2, join2)
+      self.assertAllClose(batch2, inp)
+
+  def testRGBToHSVRoundTrip(self):
+    data = [0, 5, 13, 54, 135, 226, 37, 8, 234, 90, 255, 1]
+    for nptype in [np.float32, np.float64]:
+      rgb_np = np.array(data, dtype=nptype).reshape([2, 2, 3]) / 255.
+      with self.test_session(use_gpu=True):
+        yiq = image_ops.rgb_to_yiq(rgb_np)
+        rgb = image_ops.hsv_to_rgb(yiq)
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_test.py]: I think you change this test and shouldn't.

**Review by martinwicke — COMMENTED** [review]
There are some comments, but in principle, this looks good. Are you interested in implementing YUV while you're at it? It's much more common than YIQ, and would look pretty much identical. Bonus: You could even implement a more generic kernel which applies a color transformation matrix, and make tiny python wrappers which apply the appropriate matrices for YUV and YIQ conversions.

**Comment by mystic123** [comment]
Yeah, I will fix this issues and implement YUV in the way you suggested ;)

**Comment by drpngx** [comment]
@mystic123 did the last comment resolve all comments?

**Comment by mystic123** [comment]
@drpngx Yes, it resolves all issues

**Comment by drpngx** [comment]
Thanks! OK, then it should be ready for review @martinwicke

**Comment by martinwicke** [comment]
Sorry, I should have thought of this earlier, but it appears to me that ColorTransform is a special case of `tf.tensordot` -- could you simply use that in your python conversion functions? Then we save ourselves an extra kernel to build, test, and maintain.

**Comment by mystic123** [comment]
@martinwicke sure, I thought about it as well but you suggested implementing kernel so I did it that way. No problem, I will refactor python wrappers to user tensorflow ops

**Comment by mystic123** [comment]
@drpngx @martinwicke last commit adds implementation which uses tensordot op

**Comment by asimshankar** [comment]
(Okay for API review)

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Hooray! Thank you!

**Comment by drpngx** [comment]
`image_ops_test` fails:

```
ERROR: testBatch (__main__.RGBToYUVTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmpfs/tmp/bazel/execroot/org_tensorflow/bazel-out/k8-py3-opt/bin/tensorflow/python/image_ops_test.runfiles/org_tensorflow/tensorflow/python/ops/image_ops_test.py", line 126, in testBatch
    inp = np.random.randint(0, 255, *shape).astype(nptype) / 255
  File "mtrand.pyx", line 905, in mtrand.RandomState.randint
TypeError: randint() takes at most 4 positional arguments (6 given)

```

**Comment by mystic123** [comment]
@drpngx I fixed that bug, now all tests should pass

**Comment by drpngx** [comment]
Thanks @mystic123 !

Jenkins, test this please

**Comment by drpngx** [comment]
```
======================================================================
ERROR: testBatch (__main__.RGBToYUVTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/tensorflow-pull-requests-cpu-python3/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-py3-opt/bin/bazel_pip/tensorflow/python/image_ops_test.runfiles/org_tensorflow/bazel_pip/tensorflow/python/ops/image_ops_test.py", line 126, in testBatch
    inp = np.random.rand(shape).astype(nptype)
  File "mtrand.pyx", line 1363, in mtrand.RandomState.rand
  File "mtrand.pyx", line 861, in mtrand.RandomState.random_sample
  File "mtrand.pyx", line 167, in mtrand.cont0_array
TypeError: 'tuple' object cannot be interpreted as an integer

----------------------------------------------------------------------
Ran 30 tests in 1.281s

FAILED (errors=1)
area_ratio_hist  [247 325 299 258 237 221 169 127  83  34]
================================================================================
```

**Comment by mystic123** [comment]
@drpngx now it should work

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
@mystic123 please fix the goldens:

```
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```

**Comment by mystic123** [comment]
@drpngx 
I executed that command, and everything is ok:

`
$ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens True
s.s.
----------------------------------------------------------------------
Ran 4 tests in 0.001s

OK (skipped=2)
`
`
$ git status
On branch feature/rgb_to_yiq
Your branch is up-to-date with 'origin/feature/rgb_to_yiq'.
nothing to commit, working directory clean
`

**Comment by martinwicke** [comment]
Can you run this on Mac or Linux, with Python2? The compatibility tests don't execute on Py3 (since what Py3 would return is different).

**Comment by rmlarsen** [comment]
@martinwicke it looks like api_compatibilty failures. It looks like the goldens were not updated. @mystic123 can you try updating them one more time?

**Comment by mystic123** [comment]
@drpngx @martinwicke @rmlarsen thank you for your help! :) 
I run api_compatibility_test on Python2 in docker with Ubuntu.

**Comment by rmlarsen** [comment]
@mystic123 Thanks!

**Pull request merged**